### PR TITLE
GH-88597: Added command line interface to UUID module.

### PR DESCRIPTION
**Pull request by achhina**
UUID module now supports command line usage.
```python
❯ ./python.exe -m uuid             
5f2d57b1-90e8-417c-ba5d-69b9b6f74289

❯ ./python.exe -m uuid -h          
usage: uuid.py [-h] [-u {uuid1,uuid3,uuid4,uuid5}] [-ns NAMESPACE] [-n NAME]

Generates a uuid using the selected uuid function.

options:
  -h, --help            show this help message and exit
  -u {uuid1,uuid3,uuid4,uuid5}, --uuid {uuid1,uuid3,uuid4,uuid5}
                        The function to use to generate the uuid. By default uuid4 function is used.
  -ns NAMESPACE, --namespace NAMESPACE
                        The namespace used as part of generating the uuid. Only required for uuid3/uuid5 functions.
  -n NAME, --name NAME  The name used as part of generating the uuid. Only required for uuid3/uuid5 functions.
```

- [X] Tests ran locally

* Issue: gh-88597

### CODE REVIEW

**Comment by achhina** [comment]
Hi @nanjekyejoannah, as you reviewed the first PR for this issue, would you be able to take a look at this one when you get a chance? I ended up moving the CLI logic to a separate function and importing & using argparse only in there to keep the rest of the module lightweight.

File: Lib/uuid.py [hunk 1]
```diff
@@ -42,6 +42,17 @@
     # make a UUID from a 16-byte string
     >>> uuid.UUID(bytes=x.bytes)
     UUID('00010203-0405-0607-0809-0a0b0c0d0e0f')
+
+Command Line Usage:
```
Comment (arhadthedev) [inline Lib/uuid.py]: Probably, it's better to move it into a separate section as in https://docs.python.org/3/library/ast.html#command-line-usage and https://docs.python.org/3/library/trace.html#command-line-usage.

Comment (achhina) [reply Lib/uuid.py]: Sure, I've gone ahead and moved that to its own CLI example section. Also added a corresponding section for CLI like in the docs linked.

File: Doc/library/uuid.rst [hunk 2]
```diff
@@ -261,6 +261,47 @@ of the :attr:`variant` attribute:
...
+.. program:: uuid
+
+.. cmdoption:: -h, --help
+
+   Show the help message and exit.
+
+.. cmdoption:: -u <uuid>
+               --uuid <uuid>
+
+   Specify the function to use to generate the uuid. By default :func:`uuid4`
```
Comment (sobolevn) [inline Doc/library/uuid.rst]:
```suggestion
   Specify the function name to use to generate the uuid. By default it uses :func:`uuid4`
```

File: Doc/library/uuid.rst [hunk 3]
```diff
@@ -261,6 +261,47 @@ of the :attr:`variant` attribute:
...
+               --namespace <namespace>
+
+   The namespace used as part of generating the uuid. Only required for
+   :func:`uuid3`/:func:`uuid5` functions.
+
+.. cmdoption:: -n <name>
+               --name <name>
+
+   The name used as part of generating the uuid. Only required for
+   :func:`uuid3`/:func:`uuid5` functions.
```
Comment (sobolevn) [inline Doc/library/uuid.rst]:
```suggestion
   :func:`uuid3` / :func:`uuid5` functions.
```

Just to be safe with `rst`, it breaks sometimes on things like this.

File: Doc/library/uuid.rst [hunk 4]
```diff
@@ -261,6 +261,47 @@ of the :attr:`variant` attribute:
...
+               --uuid <uuid>
+
+   Specify the function to use to generate the uuid. By default :func:`uuid4`
+   is used.
+
+.. cmdoption:: -ns <namespace>
+               --namespace <namespace>
+
+   The namespace used as part of generating the uuid. Only required for
+   :func:`uuid3`/:func:`uuid5` functions.
```
Comment (sobolevn) [inline Doc/library/uuid.rst]:
```suggestion
   :func:`uuid3` / :func:`uuid5` functions.
```

File: Doc/library/uuid.rst [hunk 5]
```diff
@@ -301,3 +342,19 @@ Here are some examples of typical usage of the :mod:`uuid` module::
...
+
+.. _uuid-cli-example:
+
+Command-Line Example
+--------------------
+
+Here are some examples of typical usage of the :mod:`uuid` command line interface:
+
+    # generate a random uuid - by default uuid4() is used
+    >>> python -m uuid
```
Comment (sobolevn) [inline Doc/library/uuid.rst]: We generally use `>>>` in python repls, I suggest replacing this with `$` not indicate that this is a raw shell.

File: Doc/library/uuid.rst [hunk 6]
```diff
@@ -301,3 +342,19 @@ Here are some examples of typical usage of the :mod:`uuid` module::
    >>> uuid.UUID(bytes=x.bytes)
    UUID('00010203-0405-0607-0809-0a0b0c0d0e0f')
 
+
+.. _uuid-cli-example:
+
+Command-Line Example
+--------------------
+
+Here are some examples of typical usage of the :mod:`uuid` command line interface:
```
Comment (sobolevn) [inline Doc/library/uuid.rst:350]:
```suggestion
Here are some examples of typical usage of the :mod:`uuid` command line interface::
```

File: Lib/uuid.py [hunk 7]
```diff
@@ -728,9 +728,58 @@ def uuid5(namespace, name):
...
+
 # The following standard UUIDs are for use with uuid3() or uuid5().
 
 NAMESPACE_DNS = UUID('6ba7b810-9dad-11d1-80b4-00c04fd430c8')
 NAMESPACE_URL = UUID('6ba7b811-9dad-11d1-80b4-00c04fd430c8')
 NAMESPACE_OID = UUID('6ba7b812-9dad-11d1-80b4-00c04fd430c8')
 NAMESPACE_X500 = UUID('6ba7b814-9dad-11d1-80b4-00c04fd430c8')
+
+if __name__ == "__main__":
+    raise SystemExit(main())
```
Comment (sobolevn) [inline Lib/uuid.py]: `main` does not return `int` objects, do we really need to use `raise` here?

**Review by sobolevn — COMMENTED** [review]
Thank you!

I know that not all CLI entrypoints are tested, but I think we should start doing it. Especially because you have logic in your `main`, not just some other function call.

File: Lib/uuid.py (hunk 7 above)
Comment (achhina) [reply Lib/uuid.py]: Ah yes, originally `main` was going to return an `int` but as it's relatively simple I ended up not bothering. Removed.

File: Doc/library/uuid.rst (hunk 5 above)
Comment (achhina) [reply Doc/library/uuid.rst]: I was getting some build errors after this - `Could not lex literal_block as "python3". Highlighting skipped.`. I tried passing it a directive right after the `::` but looked like it didn't pick it up, but using a separate code-block with a directive seems to work now.

**Comment by achhina** [comment]
Hi @sobolevn, thanks for taking a look! I believe I've addressed all the comments, and have added in some unit tests. Would you be able to take a look again when you have time?

File: Doc/library/uuid.rst [hunk 8]
```diff
@@ -261,6 +261,47 @@ of the :attr:`variant` attribute:
...
 
+.. _uuid-cli:
+
+Command-Line Usage
+------------------
+
+.. versionadded:: 3.12
+
+The :mod:`uuid` module can be executed as a script from the command line.
+It is as simple as:
```
Comment (sobolevn) [inline Doc/library/uuid.rst]:
```suggestion
The :mod:`uuid` module can be executed as a script from the command line:
```

File: Doc/library/uuid.rst [hunk 9]
```diff
@@ -261,6 +261,47 @@ of the :attr:`variant` attribute:
...
+.. program:: uuid
+
+.. cmdoption:: -h, --help
+
+   Show the help message and exit.
+
+.. cmdoption:: -u <uuid>
+               --uuid <uuid>
+
+   Specify the function name to use to generate the uuid. By default it uses :func:`uuid4`
```
Comment (sobolevn) [inline Doc/library/uuid.rst]:
```suggestion
   Specify the function name to use to generate the uuid. By default :func:`uuid4`
```

File: Lib/test/test_uuid.py [hunk 10]
```diff
@@ -675,6 +675,33 @@ def test_uuid_weakref(self):
...
+        self.assertEqual(cm.exception.code, 2)
+
+    @mock.patch.object(sys, "argv", [""])
+    def test_cli_uuid4_outputted_with_no_args(self):
+        stdout = io.StringIO()
+        with contextlib.redirect_stdout(stdout):
+            self.uuid.main()
+
+        # Output uuid should be in the format of uuid4
+        self.assertRegex(
```
Comment (sobolevn) [inline Lib/test/test_uuid.py]: I think instead we can pass the output to the `uuid.uuid4` function to be parsed. This is more clear and more correct.

Also, required params for other functions (namespace, name) are not tested. Let's add them!

File: Lib/uuid.py [hunk 11]
```diff
@@ -728,9 +728,58 @@ def uuid5(namespace, name):
...
+
 # The following standard UUIDs are for use with uuid3() or uuid5().
 
 NAMESPACE_DNS = UUID('6ba7b810-9dad-11d1-80b4-00c04fd430c8')
 NAMESPACE_URL = UUID('6ba7b811-9dad-11d1-80b4-00c04fd430c8')
 NAMESPACE_OID = UUID('6ba7b812-9dad-11d1-80b4-00c04fd430c8')
 NAMESPACE_X500 = UUID('6ba7b814-9dad-11d1-80b4-00c04fd430c8')
+
+if __name__ == "__main__":
+    SystemExit(main())
```
Comment (sobolevn) [inline Lib/uuid.py]:
```suggestion
    main()
```

**Review by sobolevn — COMMENTED** [review]
Thanks! Several more ideas / comments :)

I think this can also be added to `whatsnew/3.12`, because this feature is quite useful.

File: Lib/test/test_uuid.py (hunk 10 above)
Comment (achhina) [reply Lib/test/test_uuid.py]: Hmm, I wasn't exactly sure of how to pass the output to `uuid.uuid4`, so instead I passed it to `uuid.UUID` and then checked string equality, and if the version was `4`. Hopefully that works? Also, added similar acceptance checks for the other functions requiring `namespace` and `name`!

**Comment by achhina** [comment]
Hi @sobolevn! I've tried to address all the comments you provided, would you be able to take another look?

**Review by sobolevn — APPROVED** [review]
LGTM! But, this will require a core-dev acceptance as well.

**Review by gpshead — APPROVED** [review]

**Comment by gpshead** [comment]
thanks!

**Comment by merwok** [comment]
For next PRs to Python repos, please don’t force push, it makes reviews less nice (comments lose contexts, github notifs don’t link to a useful page, etc).  See https://devguide.python.org/getting-started/pull-request-lifecycle/index.html

**Pull request merged**